File: app/src/main/java/benor/agendawidget/MainActivity.java
package benor.agendawidget;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.os.Bundle;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.SimpleAdapter;
import android.widget.Spinner;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Objects;
import java.util.Timer;
import java.util.TimerTask;

import androidx.core.app.ActivityCompat;
import benor.MLog.MLog;

public class MainActivity extends Activity {
	public final int PERMISSIONS_REQUEST = 1;
	public final int BG_COLOR_REQUEST = 2;
	private List<Calendar> calendars;
	private List<ResolveInfo> pkgAppsList;
	private List<HashMap<String, String>> installedAppsData;
	private boolean loadedApps = false;
	private boolean displayingStartScreen;


	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		if (getApplicationContext().checkSelfPermission(Manifest.permission.WRITE_CALENDAR) != PackageManager.PERMISSION_GRANTED) {
			ActivityCompat.requestPermissions(this,
					new String[]{Manifest.permission.READ_CALENDAR, Manifest.permission.WRITE_CALENDAR},
					PERMISSIONS_REQUEST);
			return;
		}
		Globals.init(getApplicationContext());
		Globals.editListInit = false;
	}

	@Override
	protected void onActivityResult(int requestCode, int resultCode, Intent data) {
		if (requestCode == BG_COLOR_REQUEST) {
			if (data.hasExtra("color")) {
				Globals.DB.setBgColor(data.getIntExtra("color", 0));
			} else {
				Globals.DB.setBgColor(null);
			}
		} else {
			super.onActivityResult(requestCode, resultCode, data);
		}
	}

	@Override
	protected void onResume() {
		super.onResume();
		calendars = Calendar.readCalendars();

		setScreen();
		new Thread(new Runnable() {
			@Override
			public void run() {
				if (!loadedApps) {
					final Intent mainIntent = new Intent(Intent.ACTION_MAIN, null);
					mainIntent.addCategory(Intent.CATEGORY_LAUNCHER);
					PackageManager pm = Globals.con.getPackageManager();
					pkgAppsList = pm.queryIntentActivities(mainIntent, 0);

					installedAppsData = new ArrayList<>();
					for (ResolveInfo app : pkgAppsList) {
						HashMap<String, String> map = new HashMap<>();
						map.put("name", app.loadLabel(pm).toString());
						map.put("intent", app.activityInfo.packageName + "!" + app.activityInfo.name);
						installedAppsData.add(map);
					}
				}
				Collections.sort(installedAppsData, new Comparator<HashMap<String, String>>() {
					@Override
					public int compare(HashMap<String, String> item1, HashMap<String, String> item2) {
						return Objects.requireNonNull(item1.get("name")).compareTo(Objects.requireNonNull(item2.get("name")));
					}
				});
				runOnUiThread(new Runnable() {

					@Override
					public void run() {
						loadedApps = true;
						findViewById(R.id.btn_select_calendar_app).setEnabled(true);
					}
				});

			}
		}).start();
	}

	private void setScreen() {
		setContentView(R.layout.activity_main);
		displayingStartScreen = true;
		Globals.editListInit = false;
		ListView cals = this.findViewById(R.id.LV_cals);
		cals.setAdapter(new CalsAdapter(getApplicationContext(), calendars));

		handleLanguageSelection();
		findViewById(R.id.btn_select_bg_color).setOnClickListener(new OnClickListener() {
			@Override
			public void onClick(View v) {
				Intent intent = new Intent(getApplicationContext(), ColorSelector.class);
				intent.putExtra("color", Globals.DB.getBgColor());
				startActivityForResult(intent, BG_COLOR_REQUEST);
			}
		});


		findViewById(R.id.btn_select_calendar_app).setOnClickListener(new OnClickListener() {
			@Override
			public void onClick(View v) {
				String[] from = new String[]{"name"};
				int[] to = new int[]{R.id.app_name};

				ListView apps = new ListView(Globals.con);
				SimpleAdapter adapter = new SimpleAdapter(Globals.con, installedAppsData, R.layout.app_item, from, to);
				apps.setAdapter(adapter);
				apps.setOnItemClickListener(new OnItemClickListener() {

					@Override
					public void onItemClick(AdapterView<?> arg0, View arg1,
											int position, long id) {
						MLog.i("clicked " + installedAppsData.get(position).get("intent"));
						Globals.DB.setCalendarApp(installedAppsData.get(position).get("intent"));
						setScreen();
					}
				});
				displayingStartScreen = false;
				setContentView(apps);
			}
		});

		if (loadedApps) {
			findViewById(R.id.btn_select_calendar_app).setEnabled(true);
		}
		Timer t = new Timer();
		t.schedule(new TimerTask() {
			@Override
			public void run() {
				MLog.i("finished?");
				Globals.editListInit = true;
			}
		}, 1000);
	}

	private void handleLanguageSelection() {
		Spinner languageSelector = this.findViewById(R.id.languageSelector);
		ArrayAdapter<String> myAdapter = new ArrayAdapter<>(MainActivity.this,
				android.R.layout.simple_spinner_dropdown_item, new String[]{CONST.LANG_ENG, CONST.LANG_HEB});
		myAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
		languageSelector.setAdapter(myAdapter);
		languageSelector.setSelection(Globals.DB.getLang().equals(CONST.LANG_ENG) ? 0 : 1, false);
		languageSelector.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
			@Override
			public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
				if (i == 0) {
					Globals.DB.setLang(CONST.LANG_ENG);
				} else {
					Globals.DB.setLang(CONST.LANG_HEB);
				}
			}

			@Override
			public void onNothingSelected(AdapterView<?> adapterView) {

			}
		});
	}

	public void onBackPressed() {
		if (!displayingStartScreen) {
			setScreen();
		}
	}

	public void onPause() {
		super.onPause();
		MLog.i("widget is " + Globals.widget);
		Widget.updateWidget(getApplicationContext());
	}
}

File: app/src/main/java/benor/agendawidget/MainActivity.kt
package benor.agendawidget

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.ListView
import android.widget.SimpleAdapter
import android.widget.Spinner
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import benor.MLog.MLog
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class MainActivity : AppCompatActivity() {

    private var calendars: List<AgendaCalendar> = emptyList()
    private var installedAppsData: List<HashMap<String, String>> = emptyList()
    private var loadedApps = false
    private var displayingStartScreen = false

    private val colorPickerLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        if (result.resultCode == RESULT_OK) {
            val data = result.data
            if (data?.hasExtra("color") == true) {
                Globals.db.setBgColor(data.getIntExtra("color", 0))
            } else {
                Globals.db.setBgColor(null)
            }
        }
    }

    private val permissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestMultiplePermissions()
    ) { permissions ->
        // After permissions dialog, the system will call onResume() automatically.
        // If denied, onResume() will exit early at the permission check below.
        if (permissions[Manifest.permission.READ_CALENDAR] == true) {
            Globals.init(applicationContext)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (checkSelfPermission(Manifest.permission.WRITE_CALENDAR) != PackageManager.PERMISSION_GRANTED) {
            permissionLauncher.launch(
                arrayOf(Manifest.permission.READ_CALENDAR, Manifest.permission.WRITE_CALENDAR)
            )
            return
        }
        Globals.init(applicationContext)
        Globals.editListInit = false
    }

    override fun onResume() {
        super.onResume()
        if (checkSelfPermission(Manifest.permission.READ_CALENDAR) != PackageManager.PERMISSION_GRANTED) return
        // Ensure Globals is initialized even if onCreate() returned early (permissions flow)
        Globals.init(applicationContext)
        calendars = AgendaCalendar.readCalendars()
        setScreen()
        loadAppsInBackground()
    }

    private fun loadAppsInBackground() {
        if (loadedApps) return
        lifecycleScope.launch {
            val apps = withContext(Dispatchers.IO) {
                val mainIntent = Intent(Intent.ACTION_MAIN, null).apply {
                    addCategory(Intent.CATEGORY_LAUNCHER)
                }
                val pm = Globals.con.packageManager
                pm.queryIntentActivities(mainIntent, 0)
                    .map { app ->
                        hashMapOf(
                            "name" to app.loadLabel(pm).toString(),
                            "intent" to "${app.activityInfo.packageName}!${app.activityInfo.name}"
                        )
                    }
                    .sortedBy { it["name"] }
            }
            installedAppsData = apps
            loadedApps = true
            if (displayingStartScreen) {
                findViewById<View>(R.id.btn_select_calendar_app)?.isEnabled = true
            }
        }
    }

    private fun setScreen() {
        setContentView(R.layout.activity_main)
        displayingStartScreen = true
        Globals.editListInit = false

        val calsList = findViewById<ListView>(R.id.LV_cals)
        calsList.adapter = CalsAdapter(applicationContext, calendars)

        handleLanguageSelection()

        findViewById<View>(R.id.btn_select_bg_color).setOnClickListener {
            colorPickerLauncher.launch(
                Intent(applicationContext, ColorSelector::class.java).apply {
                    Globals.db.getBgColor()?.let { putExtra("color", it) }
                }
            )
        }

        findViewById<View>(R.id.btn_select_calendar_app).apply {
            isEnabled = loadedApps
            setOnClickListener {
                if (loadedApps) showAppsList()
            }
        }

        // Delay editListInit to prevent spurious checkbox callbacks during list setup
        Handler(Looper.getMainLooper()).postDelayed({
            MLog.i("list init complete")
            Globals.editListInit = true
        }, 1000)
    }

    private fun showAppsList() {
        displayingStartScreen = false
        val apps = ListView(this)
        val adapter = SimpleAdapter(
            this, installedAppsData, R.layout.app_item,
            arrayOf("name"), intArrayOf(R.id.app_name)
        )
        apps.adapter = adapter
        apps.setOnItemClickListener { _, _, position, _ ->
            MLog.i("clicked ${installedAppsData[position]["intent"]}")
            Globals.db.setCalendarApp(installedAppsData[position]["intent"] ?: "")
            setScreen()
        }
        setContentView(apps)
    }

    private fun handleLanguageSelection() {
        val spinner = findViewById<Spinner>(R.id.languageSelector)
        val adapter = ArrayAdapter(
            this,
            android.R.layout.simple_spinner_dropdown_item,
            arrayOf(CONST.LANG_ENG, CONST.LANG_HEB)
        )
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        spinner.adapter = adapter
        spinner.setSelection(if (Globals.db.getLang() == CONST.LANG_ENG) 0 else 1, false)
        spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                Globals.db.setLang(if (position == 0) CONST.LANG_ENG else CONST.LANG_HEB)
            }
            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
    }

    @Suppress("OVERRIDE_DEPRECATION")
    override fun onBackPressed() {
        if (!displayingStartScreen) {
            setScreen()
        } else {
            @Suppress("DEPRECATION")
            super.onBackPressed()
        }
    }

    override fun onPause() {
        super.onPause()
        MLog.i("widget is ${Globals.widget}")
        Widget.updateWidget(applicationContext)
    }
}
